examples/bigquery_user_info_updater_tool/bigquery_user_info_updater/tests/test_user_info_updater.py

```diff
@@ -0,0 +1,348 @@
+
+    def test_initial_update(self, project_id):
+        """Tests UserInfoUpdater ability to run an initial update.
+
+         Args:
+             project_id(str): ID of the project that holds the test BQ tables.
+
+         Returns:
+             True if test passes, else False.
+         """
+        # Load the first set of user updates to user_info_updates table.
+        # All other tables should be empty at this point.
+
+        self.load_csv_to_bq(
+            filename='test_data/user_info_updates_data_1.csv',
+            table=self.dataset_ref.table(self.user_info_updates_id)
+        )
+        # Run the UserInfoUpdater on the first set of updates.
+        test_updater = user_info_updater.UserInfoUpdater(
+            project_id,
+            self.dataset_id,
+            self.user_info_updates_id,
+            self.temp_user_info_updates_id,
+            self.user_info_final_id
+        )
+
+        # Get Queries to Run
+        update_query_creator = query_creator.QueryCreator(
+            schema_path=self.schema_path,
+            user_id_field_name='userId',
+            ingest_timestamp_field_name='ingestTimestamp',
+            project_id=project_id,
+            dataset_id=self.dataset_id,
+            updates_table_id=self.user_info_updates_id,
+            temp_updates_table_id=self.temp_user_info_updates_id,
+            final_table_id=self.user_info_final_id
+        )
+
+        gather_updates_query = update_query_creator.create_gather_updates_query()
+        test_updater.gather_updates(gather_updates_query)
+        merge_udpates_query = update_query_creator.create_merge_query()
+        test_updater.merge_updates(merge_udpates_query)
+
+        # Query the temp table to test that the gather_updates() function worked
+        temp_table_query_config = bigquery.QueryJobConfig()
+        temp_table_query_config.use_legacy_sql = False
+        temp_table_query = self.bq_client.query(
+            query='SELECT * FROM `{0:s}.{1:s}.{2:s}`'.format(
+                project_id,
+                self.dataset_id,
+                self.temp_user_info_updates_id
+            ),
+            job_config=temp_table_query_config,
+            location='US'
+        )
+        temp_table_query.result()
+        temp_table_results_df = temp_table_query.to_dataframe()\
+            .sort_values(by=['userId']).reset_index(drop=True)
+
+        # Gather expected results for comparison
+        abs_path = os.path.abspath(os.path.dirname(__file__))
+        expected_temp_data_file = os.path.join(
+            abs_path,
+            'test_data/temp_user_info_updates_expected_1.csv'
+        )
+        expected_temp_table_df = pd.read_csv(expected_temp_data_file)
+        # convert ingestTimestamp to datetime
+        expected_temp_table_df['ingestTimestamp'] = pd.to_datetime(
+            expected_temp_table_df['ingestTimestamp']
+        )
+
+        # Compare results
+        pd.testing.assert_frame_equal(
```

Comment:
jaketf on Aug 27, 2019 Member
pandas provides a method pandas.util.testing.assert_frame_equal for this which will raise more informative error messages.